core/src/state.hpp

```diff
@@ -152,8 +152,18 @@ class state
     void imx_deregister_callback();
 
     void imx_callback(uint32_t imx_id);
-};
 
+    // This is intended to be used to take the actions given in the actions
+    // parameter, and load them into the _actions member of this class. Used by
+    // keyboard processors to set the output actions, and is a long-term
+    // replacement for the actions()::push_*() functions. Note that the
```

Comment:
I have a different understanding. 
We have `engines` oneside of the core and `keyboard processors` on the other side. On the `engine` side we need `km_core_state_get_actions` to obtain the `km_core_actions` struct for `engines` who consume it. On the keyboard processor side we need `set_actions` to do the reverse to handle keyboard processors that produce a `km_core_actions` struct. So that we can convert the actions struct back into a km_core_action_items` list for the `engines` that still consume this. 
The keyboard processors will do a call something like `km_core_state_set_actions` with the action struct then internally the core in that method (`km_core_state_set_actions`) will call `set_actions`. 

Current situation
1. kmx keyboard processor does not have an understanding of the `km_core_actions`. It uses the `actions()::push_*()` completed with `actions()::commit`. To create a `km_core_action_items` list.  This will be converted to `km_core_actions` struct when a core consumer (engine) calls `km_core_state_get_actions`. MacOS engine is on this path already.
2. In implementing the LDML processor transformations it was discovered it needed and beneficial to be able to produce `km_core_actions` struct and be able to set it in the core. 
3. As the Windows and Linux Engines are still consuming`km_core_action_items` list they will need this PR with `set_actions` to convert the `km_core_actions` struct back to the `km_core_action_items` 
4. If all platform engines are updated to consume `km_core_actions` struct. Then this `set_actions` method will no longer be needed, excepted to serve the deprecated API.  Even if the kmx_processor is still pushing action_items, as the `km_core_state_get_actions` will supply the struct to the engine.

Reply:
I think that's what my code comment says, isn't it?

Reply:
//Used by
    // keyboard processors to set the output actions,
The keyboard processors should not be using `set_actions` to set the output actions. They shouldn't know it exists. They use the push/set call to get the `km_core_actions` to the core. The core can internally at this time call `set_actions` to satisfy the engines that need the `km_core_action_items` via the to-be-depricated `km_core_state_action_items` call.
// and is a long-term
 // replacement for the actions()::push_*() functions
`set_actions` is internal to core helper method and will only live for as long as the deprecated actions list is supported. The replacement for the `actions()::push_*() functions` will be a method that just pushes the a `km_core_actions` struct to the core. It could mirror the get `km_core_state_set_actions` or it could match the current methods `actions()::push_actions()

That is my understanding anyway.

Reply:
Oh, I think the penny has finally dropped. I was seeing this `set_actions` just as the conversion function from actions struct to actions list. Something that the method that actually sets the actions calls/runs. That way we could keep what is the main functionality and which is assisting the deprecated code separate. 
I presume then that this `set_actions` is not complete in that in the future it will store the struct in a yet-to-be-added private member of "_actions_struct".

Reply:
so this is a long term function, not temporary.

Reply:
> I presume then that this `set_actions` is not complete in that in the future it will store the struct in a yet-to-be-added private member of "_actions_struct".

Eventually all the `push_` methods will disappear, and `set_actions` will be the only way for a keyboard processor to set the actions; this will mean some changes internally for `km::core::state` and `km::core::actions` objects, yes.